src/useUnloadConfirmation.ts

```diff
@@ -0,0 +1,80 @@
+import { MutableRefObject, useEffect, useRef } from 'react'
+import { router } from './router'
+
+interface UnloadConfig {
+  unloadMessage: string | null
+  onConfirm: (() => void) | undefined
+}
+
+/**
+ * This makes sure that unloading the page requires user confirmation - if unloadMessage is set.
+ * Uses the browser's native `beforeunload` prevention feature.
+ *
+ * Additionally, it stores a list of "hooks" for the router to check via the `preventUnload` method when navigating
+ */
+export function useUnloadConfirmation(unloadMessage: string | null, onConfirm?: () => void): void {
+  const routerUnloadFunctionRef = useRef({
+    unloadMessage,
+    onConfirm,
+  })
+
+  router.cache.__unloadConfirmations = (router.cache.__unloadConfirmations || []) as MutableRefObject<UnloadConfig>[]
+
+  useEffect(() => {
+    const unmountFunctions: (() => void)[] = []
+
+    // Native browser unloading (e.g. page refresh)
+    if (unloadMessage) {
+      const beforeUnloadHandler = (e: BeforeUnloadEvent): void => {
+        // Cancel the event to show unsaved changes dialog
+        e.preventDefault()
+        e.returnValue = ''
+      }
+
+      window.addEventListener('beforeunload', beforeUnloadHandler)
+      unmountFunctions.push(() => window.removeEventListener('beforeunload', beforeUnloadHandler))
+    }
+
+    // Kea-router based unloading (e.g. push(), replace())
+    if (unloadMessage) {
+      routerUnloadFunctionRef.current = {
+        unloadMessage,
+        onConfirm,
```

Comment:
This is one place where the new and unreleased [useEvent API](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md) would come in handy. As it is, `onConfirm` can't change independently of `unloadMessage`. This might be unexpected to some.

And obviously we don't want to run this `useEffect` every time this function changes, because it could be an inline function that changes every time it's rendered...

A hack is to write `routerUnloadFunctionRef.current.onConfirm = onConfirm` directly in the function body before this `useEffect`. This might be enough.

Reply:
Yeah I had the same thoughts and I think that hack you suggest solves the issue well enough 👍